File: exercises_3/matlab/solutions/exampleSolutionPart2.m
clear;
close all;

%load lung MR images
cine_MR_1 = imread('cine_MR_1.png');
cine_MR_2 = imread('cine_MR_2.png');
cine_MR_3 = imread('cine_MR_3.png');

%convert to double and change to 'standard orientation'
cine_MR_1 = flip(double(cine_MR_1'), 2);
cine_MR_2 = flip(double(cine_MR_2'), 2);
cine_MR_3 = flip(double(cine_MR_3'), 2);


demonsScript

figure;
dispImage(target_full - trans_image);
figure;
dispColourOverlayImage(target_full, trans_image);

J = calcJacobian(def_field);
figure
dispImage(J);
colorbar;
figure
dispImage(J<=0);
disp(sum(J(:)<=0));


diffeoDemonsScript
J = calcJacobian(def_field);
disp(sum(J(:)<=0));
figure
dispImage(J);

File: exercises_3/python/solutions/exampleSolutionPart2.py
import skimage.io
import numpy as np
import utils
import matplotlib.pyplot as plt
import demons

img1 = skimage.io.imread('cine_MR_1.png')
img2 = skimage.io.imread('cine_MR_2.png')
img3 = skimage.io.imread('cine_MR_3.png')

img1 = (np.flipud(np.double(img1))).T
img2 = (np.flipud(np.double(img2))).T
img3 = (np.flipud(np.double(img3))).T


plt.close('all')
reg_img, def_field = demons.runRegistration(img1, img2, sigma_elastic=0.5, sigma_fluid=1, num_lev=3)
plt.close('all')
plt.figure()
utils.dispImage(img1 - reg_img)
plt.figure()
utils.dispColourOverlayImage(img1, reg_img)
plt.figure()
utils.dispDefField(def_field, spacing=2)

J, J_Mat = utils.calcJacobian(def_field)
plt.figure()
utils.dispImage(J)
plt.colorbar()
plt.figure()
folding = J <= 0
utils.dispImage(folding)
print(np.sum(folding))